src/python/pants/jvm/dependency_inference/artifact_mapper.py

```diff
@@ -56,10 +57,18 @@ def addresses_for_coordinates(
 
 
 class MutableTrieNode:
+    __slots__ = [
+        "children",
+        "recursive",
+        "addresses",
+        "first_party",
+    ]  # don't use a `dict` to store attrs
+
     def __init__(self):
         self.children: dict[str, MutableTrieNode] = {}
         self.recursive: bool = False
         self.addresses: OrderedSet[Address] = OrderedSet()
+        self.first_party: bool = False
```

Comment:
If we end up having the `!` syntax for `packages`, we may want to redefine this field for more complicated exclusions, but for now, this will do.

Reply:
Agreed.